context.go

```diff
@@ -2,107 +2,154 @@ package parens
 
 import (
 	"errors"
+	"fmt"
+	"sync"
 
 	"github.com/spy16/parens/value"
 )
 
-var _ Context = (*basicContext)(nil)
-
-const globalFrame = "<global>"
-
 // ErrNotFound is returned by Evaluator when a symbol resolution fails.
 var ErrNotFound = errors.New("not found")
 
-// Context encapsulates execution state.  It is passed to Evaluator in order to evaluate
-// a form within a given execution context.  The default Context implementation IS NOT
-// safe for concurrent use without external synchronization.
-type Context interface {
-	Push(StackFrame)
-	Pop() StackFrame
+// ConcurrentMap to store vars in a stack frame.
+type ConcurrentMap interface {
+	Load(string) (value.Any, bool)
+	Store(string, value.Any)
 
-	SetGlobal(name string, val value.Any)
-	Resolve(name string) (value.Any, error)
-
-	NewChild() Context
+	// Native map[string]value.Any.  Useful for ranging over values, e.g. when
+	// printing a stack frame.
+	Native() map[string]value.Any
 }
 
-// NewContext returns a global context using the default Context implementation.
-// The returned Context IS NOT safe for concurrent use without external synchronization.
-func NewContext() Context {
-	return newContext([]StackFrame{{Name: globalFrame}})
-}
+// MapFactory is a constructor for an implementation of ConcurrentMap.
+type MapFactory func() ConcurrentMap
 
 // StackFrame .
 type StackFrame struct {
```

Comment:
Do we need to expose this? Only InvokeExpr would manipulate stack i think?.. If we need that flexibility, what if we just made an interface for stack ?